File: Ch01/ex1_05.rb
#!/usr/bin/env ruby
# ex1_05.rb

=begin rdoc
Kevin C. Baird
SICP in Ruby
Test applicative-order ("evaluate args and then apply")
vs. 
normal-order ("fully expand and then reduce")
evaluation.
=end

RECIP = lambda { |x| 1.0/x }
IDENT = lambda { |x| x     }

def test_order(a, the_proc)
  return nil if a.zero?
  the_proc[a]
end

=begin
Applicative order exits early with the nil return guard, never calling SOURCE.
Normal order fully expands the call to SOURCE and then returns its value.
=end

File: Ch01/Ex1_05.hs
{-
 - Ex1_05.hs
 - Kevin C. Baird
 - SICP in Haskell
 -
 - Test applicative-order ("evaluate args and then apply")
 - vs.
 - normal-order ("fully expand and then reduce")
 - evaluation.
 -
 - The closest equivalent in Haskell is probably lazy evaluation.
 -}

ones :: [Int]
ones = 1 : ones

testOrder :: Int -> [Int] -> Int
testOrder x y
  | x == 0    = 0
  | otherwise = sum y

{-
 - Evaluating "testOrder 0 ones" in hugs will correctly
 - return 0, while any other x argument value will force an
 - attempted evaluation of the inifinite stream ('ones').
 -
 -}
